### PR TITLE
Make txArriveTimeout Configurable w/ CLI Flag

### PR DESCRIPTION
**Pull request by thogard785**
# Description

Following up on:
1. [PR 627 - txArriveTimeout change](https://github.com/maticnetwork/bor/pull/627)
2. [PR 707 - txArriveTimeout change reverted](https://github.com/maticnetwork/bor/pull/707) 

- Renamed `txArriveTimeout` to `txArrivalWait` to more accurately describe the purpose of the argument - this is a variable that controls how long to wait before requesting a transaction, there's no action/request/response that has been 'timed out'.
- Added a CLI flag  `txarrivalwait` to allow node operators to set their own maximum wait period before explicitly requesting an announced transaction.
- Set the default `txarrivalwait` period to 500ms (EDIT: updated from 100ms per comment from jekamas).  While the motives of the author of [PR 627](https://github.com/maticnetwork/bor/pull/627) are unclear, his analysis and data collection are very thorough.  _Extremely_ thorough for somebody's first PR...

I suspect that the optimal `txarrivalwait` will vary per node - some nodes may prioritize a decreased load or may be in higher-latency locations while others may want a very rapid response.  By allowing the setting to be configurable, we give node operators the chance to choose a setting that best fits their need.  

Note that this also will allow validators the ability to customize their interactions with the FastLane MEV system, as referenced in [PR 707](https://github.com/maticnetwork/bor/pull/707).  

# Changes

- [x] New feature (non-breaking change that adds functionality)

# Nodes audience

This 

# Checklist

- [x] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

## Testing

- [x] I have updated the tx_fetcher unit test
- [x] tested on live w/ various flag values
- [x] tested on live w/o flag (default value)

# Additional comments

- I do not have high conviction either for or against the default value of 100ms and am open to changing it. The rationale for the selection of 100ms as default is solely due to [PR 627](https://github.com/maticnetwork/bor/pull/627) being accepted. EDIT: changed to 500ms per comments from JekaMas. 

- EDIT: Added a minimum `txArrivalWait` value Added a check that sets the `f.txArrivalWait` value to the `txGatherSlack` value if `f.txArrivalWait` < `txGatherSlack`.  

The rationale is due to the context of the usage of `txGatherSlack` by the fetcher:
`if time.Duration(f.clock.Now()-instance)+txGatherSlack > f.txArrivalWait {...` 
(found on lines [451](https://github.com/maticnetwork/bor/pull/734/files#diff-e990a3664cf221db4c72bc6c267ef9b7a222fc77e91cc87af415201915bdf1c6R451) and [488](https://github.com/maticnetwork/bor/pull/734/files#diff-e990a3664cf221db4c72bc6c267ef9b7a222fc77e91cc87af415201915bdf1c6R488) )

if `f.txArrivalWait` is less than `txGatherSlack`, it is impossible for the above conditional to return false, rendering the value of `f.txArrivalWait` meaningless and potentially leading to node operators misunderstanding the impact of the `txArrivalWait` value they set. 

(The exception to the above is if the node operator has modified the hardcoded `gatherSlack` variable (default: 100ms) to be less than the `txGatherSlack` variable (default: 20ms).  Comparing `f.txArrivalWait` against these two values is `txArrivalWait`'s only hot path usage in bor. 

[d56370b](https://github.com/thogard785/bor/commit/d56370b9101717d7b48c5862416a8069df949174)

### CODE REVIEW

File: docs/cli/server.md
Comment (JekaMas) [inline docs/cli/server.md:149]: let's keep it 500 by far.

File: internal/cli/server/config.go
Comment (JekaMas) [inline internal/cli/server/config.go:136]: could you check if it is possible to use time.Duration here?

File: p2p/server.go
Comment (JekaMas) [inline p2p/server.go:162]: the same as before with time.Duration type

**Review by JekaMas — APPROVED** [review]
LGTM with few comments.

File: docs/cli/server.md
Comment (thogard785) [reply docs/cli/server.md]: updated default to 500:

commit [b5ff754b7](https://github.com/thogard785/bor/commit/b5ff754b70c15fa0955173b3d6ea660bce3b1fa2)

File: p2p/server.go
Comment (thogard785) [reply p2p/server.go]: updated to time.Duration

commit [3c23de2](https://github.com/thogard785/bor/commit/3c23de2486a3eaee9f938aef883ee26c551c1dee)

File: docs/cli/server.md
Comment (JekaMas) [inline docs/cli/server.md:149]: 500ms

File: internal/cli/server/config.go
Comment (JekaMas) [inline internal/cli/server/config.go:137]: could you check how it works with and without provided by user value, especially `optional` tag?

File: docs/cli/server.md
Comment (thogard785) [reply docs/cli/server.md]: updated

[cfb9eb3](https://github.com/thogard785/bor/commit/cfb9eb3d1991d5ded0fb4b26b5c3ee096947161f)

File: internal/cli/server/config.go
Comment (thogard785) [reply internal/cli/server/config.go:137]: added a log line to print the `txArrivalWait` value inside of the start of the fetcher's go routine

[b8be0da](https://github.com/thogard785/bor/commit/b8be0dace866c6ca1995eb2f344c2beaf0790d5c)

can confirm that no flag being used results in a value of 500ms being used by fetcher. 

can confirm that a nonsensical value for the flag results in bor not launching
`invalid value "purple" for flag -txarrivalwait: parse error` (then prints the acceptable commands)

**Comment by thogard785** [comment]
Added a check that sets the txArrivalWait value to the txGatherSlack value if txArrivalWait < txGatherSlack.  

The rationale is due to the context of the usage of txGatherSlack by the fetcher:
`if time.Duration(f.clock.Now()-instance)+txGatherSlack > f.txArrivalWait {...` 
(found on lines [451](https://github.com/maticnetwork/bor/pull/734/files#diff-e990a3664cf221db4c72bc6c267ef9b7a222fc77e91cc87af415201915bdf1c6R451) and [488](https://github.com/maticnetwork/bor/pull/734/files#diff-e990a3664cf221db4c72bc6c267ef9b7a222fc77e91cc87af415201915bdf1c6R488) )

if `f.txArrivalWait` is less than `txGatherSlack`, it is impossible for this bool to resolve false, rendering the value of txArrivalWait meaningless and potentially leading to node operators misunderstanding the impact of the ``txArrivalWait`` value they set. 

(The exception to the above is if the node operator has modified the hardcoded `gatherSlack` variable (default: 100ms) to be less than the `txGatherSlack` variable (default: 20ms).  Comparing txArrivalWait against these two values is txArrivalWait's only hot path usage in bor. 

[d56370b](https://github.com/thogard785/bor/commit/d56370b9101717d7b48c5862416a8069df949174)

**Comment by thogard785** [comment]
@JekaMas Wanted to make sure you saw the change to add in a min value for the `f.txArrivalWait` - I think that commit went through right as you hit approve and so I wanted to make sure nothing slipped through the cracks.

 @0xsharma  @temaniarpit27

File: docs/cli/server.md
Comment (JekaMas) [inline docs/cli/server.md]: Sorry for nitpicking: could you chenge 500 to 500ms?

File: cmd/utils/flags.go
Comment (JekaMas) [inline cmd/utils/flags.go]: I believe not it is not about milliseconds. Could you change comments accordingly?

File: docs/cli/server.md
Comment (JekaMas) [inline docs/cli/server.md]: Should we change this comment also: 500->500ms?

File: cmd/utils/flags.go
Comment (thogard785) [reply cmd/utils/flags.go]: this is odd - I had already updated that line. I'm not sure why it's pulling up an older commit's text..

https://github.com/thogard785/bor/blob/d56370b9101717d7b48c5862416a8069df949174/cmd/utils/flags.go#L761

GitHub is out to get me I swear lol.

File: docs/cli/server.md
Comment (thogard785) [reply docs/cli/server.md]: This is odd... this one is also pulling up an older commit somehow.  on my end it looks updated:

https://github.com/thogard785/bor/blame/fetch-timeout-config/docs/cli/server.md#L149

I'll try to figure out why the most recent commit is being overwritten. Very strange.

Comment (Elyx0) [reply docs/cli/server.md]: It looks good in https://github.com/maticnetwork/bor/pull/734/files#diff-ca9f0d80fbfe0fad5fdb2bf71db6b22dada4751d4107ac0ca8566cbd5c7c1639R149 the file changes section

Comment (thogard785) [reply docs/cli/server.md]: commit that updates this line to duration:

3c23de2

**Comment by thogard785** [comment]
@JekaMas Hey - I believe I erroneously removed @temaniarpit27 and @0xsharma from the "review" position - could you re-add them?

File: internal/cli/server/config.go
Comment (pratikspatil024) [inline internal/cli/server/config.go:137]: I think you will have to add a `Raw string` field here and give the `txarrivalwait` tag to it, and then convert that string to time.Duration.

Have a look at [this](https://github.com/maticnetwork/bor/commit/fcb722e4bf5bdda23d4c91da40ab95c9023a80c3) for reference.

**Review by pratikspatil024 — COMMENTED** [review]
Also, update all the TOML configuration files for consistency.

File: internal/cli/server/config.go
Comment (thogard785) [reply internal/cli/server/config.go:137]: I'm not sure I understand the rationale for this.  time.Duration can unmarshal from an hcl and a toml, and there are already multiple other time.Durations in the config structs.  Is there a particular reason why we'd want to use a different unmarshal format for this variable than for the other duration variables in the config?

Comment (thogard785) [reply internal/cli/server/config.go]: @pratikspatil024 

I added both - thanks for the feedback. 

d56370b

by the way, that pr you linked is something I spend a lot of time thinking about.  we should connect sometime soon and discuss.  That PR opens up a lot of doors that are currently closed and would have a lot of neat secondary effects and enable lots of neat protocols.

Comment (pratikspatil024) [reply internal/cli/server/config.go]: Sure, let's move the conversation over [there](https://github.com/maticnetwork/bor/commit/fcb722e4bf5bdda23d4c91da40ab95c9023a80c3).

**Comment by thogard785** [comment]
@pratikspatil024 hey - what are the next steps for approval on this?

**Comment by pratikspatil024** [comment]
> @pratikspatil024 hey - what are the next steps for approval on this?

Hi, @thogard785 we are currently keeping all the `MEV` labelled PRs and issues on hold. Once a few developers have some bandwidth, they will pick this up, and do some analysis.

**Comment by thogard785** [comment]
> 

sounds good!

**Comment by ajb** [comment]
> Set the default txarrivalwait period to 500ms (EDIT: updated from 100ms per comment from jekamas). While the motives of the author of https://github.com/maticnetwork/bor/pull/627 are unclear, his analysis and data collection are very thorough. Extremely thorough for somebody's first PR...

How about a conflict of interest disclosure from Fastlane labs before you go accusing others of having ulterior motives?

Who are the investors in Fastlane, and are any of them related to Polygon?

**Comment by JekaMas** [comment]
> How about a conflict of interest disclosure from Fastlane labs before you go accusing others of having ulterior motives?
> 
> Who are the investors in Fastlane, and are any of them related to Polygon?

Thank you for your interest @ajb ! Polygon dev team welcomes all opinions and proposals.
If you believe that the network will benefit from setting the timeout default value to some particular value, we are more than just interested in that kind of research.
By the time we decided to make a few constants as a params to simplify mev related research and put on hold the rest of the tasks related to mev. We are creating a mev study core dev group with focus on having as less spam as possible. If you'd like to participate and can help with this research I'd ve happy to invite you to the group.

**Comment by JekaMas** [comment]
@pratikspatil024 @temaniarpit27 @0xsharma I am ok with that change, it does nothing with the network, although will simply the research. Let's merge it, if you are ok too.

**Comment by JekaMas** [comment]
> Sure, I will participate. Please let me know.

@ajb great! By far could you give as the way to reproduce your experi.ents and data? Like what steps, setup, etc.

**Comment by ajb** [comment]
> @ajb great! By far could you give as the way to reproduce your experiments and data? Like what steps, setup, etc.

What do you mean? PR #627 is not mine.

I would be happy to do some experiments regarding the PR I authored (#292) but I'm unsure how to really advocate for that change, since with these types of changes, you would need to make the change on the network *first*, then wait for bots to change their behavior accordingly.

**Comment by asbansal** [comment]
A slightly different take - Ethereum had spam issues too although not as much as the low fee chains. Introduction of mev-geth eliminated most of that. [mev-bor](https://github.com/marlinprotocol/mev-bor) as an alternative to bor already exists and it can be updated to feature full block templates like mev-boost. Allowing client diversity may be a simpler solution to the spam issue. 

@JekaMas we would be interested in joining the group as well!

**Comment by temaniarpit27** [comment]
@thogard785 we would like to go ahead with your PR. Can you please pull in the latest changes and resolve conflicts?

**Comment by CitizenSearcher** [comment]
I had a concern regarding this PR that I think could be solved with a minor addition to this PR.

There's a danger of misconfiguration of this parameter by a validator. Validator nodes rely on the TxFetcher in order to receive all transactions, since the small number of peers means most transaction will not be directly propagated. If a validator were to inadvertently set this value too high, it could result in a validator being unable to process most, or all of the transactions during their sprint. In the case of a normal validator, a small minority of transaction might still make it through with direct propagation, and with PFL validators only those transactions that win an PFL auction might make it through. This would result in a pretty bad degradation in user experience, and also could be difficult to debug by either the validator or polygon team. It's hard to predict what kind of effects on the network there long periods of no transaction processing would have on the network, so it's certainly possible this would have network level effects on other participants.

Could I suggest putting a cap on this value enforced inside bor? I think the old value of .5 seconds would make a lot of sense. I can understand why a node operator might want to lower the value, but it seems unlikely that a they would want to place set this value higher than the old default. If there's a concern over resource load, I'd suggest other settings to be considered, as this parameter would have almost no effect on resources given how little the TxFetcher is used by nodes with any reasonable number of peers.

In the PFL [whitepaper](https://www.fastlane.finance/PFL_WHITE_PAPER_1_5.pdf), the number one design objective is that the system "Incurs no transaction delay beyond the limit of what is already possible.". The suggested value would certainly help to achieve this. There might be some perverse incentive for PFL / validators to want to delay transactions with this parameter in order to extract more MEV; this kind of thing being supported opens up a lot of other questions / issues that I would need to be clearfully considered.

**Comment by thogard785** [comment]
@CitizenSearcher that's good feedback, and you're right about our objective of not delaying regular users past the "default" of 500ms.  I'm 100% ok with a 500ms ceiling.

**Comment by temaniarpit27** [comment]
@CitizenSearcher thanks a lot for the feedback
@thogard785 I think we should add a upper cap on this with 500ms as the val

**Comment by thogard785** [comment]
> @CitizenSearcher thanks a lot for the feedback
> @thogard785 I think we should add a upper cap on this with 500ms as the val
added:
a5b49c313

**Comment by thogard785** [comment]
> @thogard785 we would like to go ahead with your PR. Can you please pull in the latest changes and resolve conflicts?

@temaniarpit27 I updated the branch to resolve conflicts:
https://github.com/thogard785/bor/commit/1e926cfad9db5903740cdc587c35e89592c3daaa

File: eth/fetcher/tx_fetcher.go
Comment (temaniarpit27) [inline eth/fetcher/tx_fetcher.go:367]: i think this might flood the logs

**Review by temaniarpit27 — APPROVED** [review]

**Comment by temaniarpit27** [comment]
@thogard785 can you please fix the failing CI?

**Comment by thogard785** [comment]
> @thogard785 can you please fix the failing CI?

Hey @temaniarpit27 - sorry for the delay on this. I was sidelined for health reasons, but I'm back to 100% now. 

These changes should fix the cuddling issues the linter was having w/ the codebase.  

7d765e03b04fe7c5997e526dc78593a01c6d665c
and
63f79e355

The core/tx_list.go change seemed to be totally unrelated to what i was working on. It looks like that cuddle issue was fixed separately in the develop branch.. so i overwrote my own fix w/ the one from develop branch when i updated this branch w/ the changes to develop. 

e62cf51c45eea8305b0e85556f9e0201f0fa250b

**Comment by thogard785** [comment]
Looks like merging w/ dev again added some new checks to fail - will fix on Monday.

**Comment by temaniarpit27** [comment]
> Looks like merging w/ dev again added some new checks to fail - will fix on Monday.

Ya will need to fix testcases as well

**Comment by thogard785** [comment]
62d8cc02d32e8250308246e501774016b76784cb

updated the fuzzer.  

@temaniarpit27 codecov/project is still failing but all the other tests are passing. I'm unsure how to resolve the codecov/project fail - any advice?

**Pull request merged**
